Run search engine experiments in Vespa from python: Three ways to get started with pyvespa (#73)
vespa-engine/blog: _posts/2021-03-12-run-search-engine-experiments-in-Vespa-from-python.md

date: '2021-03-12'
tags: []
author: thigm85
image: assets/2021-03-12-run-search-engine-experiments-in-Vespa-from-python/figure_1.jpg
skipimage: true
excerpt: Three ways to get started with pyvespa.
---

**Three ways to get started with pyvespa.**

[pyvespa](https://pyvespa.readthedocs.io/en/latest/index.html) provides a python API to Vespa. The library’s primary goal is to allow for faster prototyping and facilitate Machine Learning experiments for Vespa applications. 

There are three ways you can get value out of `pyvespa`: 

1. You can connect to a running Vespa application.

2. You can build and deploy a Vespa application using pyvespa API.

3. You can deploy an application from Vespa config files stored on disk.

We will review each of those methods.

![Decorative image](/assets/2021-03-12-run-search-engine-experiments-in-Vespa-from-python/figure_1.jpg)
<p class="image-credit">Photo by <a href="https://unsplash.com/@vonshnauzer?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Egor Myznik</a> on <a href="/s/photos/three-vespa?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Unsplash</a></p>

##  Connect to a running Vespa application

> In case you already have a Vespa application running somewhere, you can directly instantiate the [Vespa](https://pyvespa.readthedocs.io/en/latest/reference-api.html#vespa.application.Vespa) class with the appropriate end-point. The example below connects to the [cord19.vespa.ai](https://cord19.vespa.ai/) application:

Reviewer: ```suggestion
In case you already have a Vespa application running somewhere, you can directly instantiate the [Vespa](https://pyvespa.readthedocs.io/en/latest/reference-api.html#vespa.application.Vespa) class with the appropriate endpoint. The example below connects to the [cord19.vespa.ai](https://cord19.vespa.ai/) application:
```

```python
from vespa.application import Vespa

app = Vespa(url = "https://api.cord19.vespa.ai")
```

> We are then good to go and ready to interact with the application through `pyvespa`, e.g., to query:

Reviewer: ```suggestion
We are then good to go and ready to interact with the application through `pyvespa`:
```

If your application requires functionality or fine-tuning not available in `pyvespa`, you simply build it directly through Vespa configuration files as shown in [many examples](https://docs.vespa.ai/en/getting-started.html) on Vespa docs. But even in this case, you can still get value out of `pyvespa` by deploying it from python based on the Vespa configuration files stored on disk. To show that, we can clone and deploy the news search app covered in this [Vespa tutorial](https://docs.vespa.ai/en/tutorials/news-3-searching.html):

```python
!git clone https://github.com/vespa-engine/sample-apps.git
```

The Vespa configuration files of the news search app are stored in the `sample-apps/news/app-3-searching/` folder:

```python
!tree sample-apps/news/app-3-searching/
```

    [01;34msample-apps/news/app-3-searching/[00m
    ├── hosts.xml
    ├── [01;34mschemas[00m
    │   └── news.sd
    └── services.xml
    
    1 directory, 3 files

We can then deploy to a Docker container from disk:

```python
from vespa.package import VespaDocker

vespa_docker_news = VespaDocker(

>     disk_folder="/Users/username/sample-apps/news/app-3-searching/", # Docker require absolute path

Reviewer: ```suggestion
    disk_folder="/Users/username/sample-apps/news/app-3-searching/", # Docker requires absolute path
```

    container_memory="8G", 
    port=8081
)
app = vespa_docker_news.deploy_from_disk(application_name="news")
```

    Waiting for configuration server.
    Waiting for configuration server.
    Waiting for configuration server.
    Waiting for configuration server.
    Waiting for configuration server.
    Waiting for application status.
    Waiting for application status.
    Finished deployment.

Again, `app` holds an instance of the Vespa class just like our first example, and we can use it to feed and query the application just deployed. 

## Final thoughts

> We covered three different ways to connect to a `Vespa` application from python using the `pyvespa` library. Those methods provide great workflow flexibility. They allow you to quickly get started with pyvespa experimentation while enabling you to modify Vespa config files to include features not available in the pyvespa API without losing the ability to experiment with the added features.

Reviewer: ```suggestion
We covered three different ways to connect to a `Vespa` application from python using the `pyvespa` library. Those methods provide great workflow flexibility. They allow you to quickly get started with pyvespa experimentation, while enabling you to modify Vespa config files to include features not available in the pyvespa API - this without losing the ability to experiment with the added features.
```
